x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/ConsistentCache.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.security.support;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.common.cache.Cache;
+import org.elasticsearch.common.util.concurrent.ReleasableLock;
+
+import java.util.Collection;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * A wrapper of {@link Cache} that minimizes the possibility of caching stale results.
+ */
+public class ConsistentCache<K, V> {
```

Comment:
I like this version. I will update the code like it except that I might change the class name to something like `StaleMinimizingCacheWrapper` to be more meaningful and specific. 

I agree that the ConsistentCache is not a good abstraction in that its interface is unintuitive to use. The background is that I started from a very complex version where I tried to provide an unified interface for security related caches and hide all implementation details, which turned out to be too much to hide. I have since tried to decompose it to a simpler version, i.e. the ConsistentCache. The general approach for simplication is to make inner steps explicit instead of trying to hide them (this is also Tim's suggestion) and also cater for fewer use cases. But my thoughts were still biased by the initial complex version, which prevented me to further simplfy it. For this, I think "pulling the counter out explicity" is indeed the right move. So the overall journey has been from an attempt to have "one cache to rule them all" to a more pragmatic one that is just for code reuse. It is less satisfying that I was not able to achieve the initial goal. But given what we have, I think the simpler abstraction is the right one for the time being.

Reply:
Not to dwell too much on it, but `StaleMinimizing...` is not what the class does but what it is used for. You're tying the class to a particular circumstance. It is "stale minimizing" only in the particular context that it is used in. I think it would be much simpler to explain how it achieves the minimization in a comment at the use site, like in the `CompositeRolesStore` class, than in the class's javadoc. The name is also a bit intimidating, IMO.

My name is inspired by `CacheIteratorHelper` .

Reply:
One main reason that I don't think `CacheInvalidationHelper` is a proper name is because the class is essentially a `Cache`. If you look past the counter and lock, it is just a regular Cache. So I'd much prefer to call it something along the line of `XxxCache` or `XxxCacheWrapper`. "Minimizing the chane of stale entries" is the new feature added onto the exiting `Cache` class. Often times, classes like this are named as `AdjectiveNoun`, e.g. `Realm` and `CachingRealm`. So I think `StaleMinimizingCache[Wrapper]` is an accurate name. If you feel strong about `StaleMinimizing`, how about `InvalidationCountingCache[Wrapper]`? It is still accurate, but less comprehensive.